### PR TITLE
Fix backwards compatibility when backend mode isn't set

### PR DESCRIPTION
**Pull request by WarpRat**
This will address the issue opened here: https://github.com/kubernetes/kops/issues/9750

I changed the test to confirm that the gotemplate or function worked as expected. I ran the test multiple times locally with a different mix of configurations to confirm that this somewhat ugly chain of if/or/not works as intended and will address the issue.

### CODE REVIEW

File: upup/models/cloudup/resources/addons/authentication.aws/k8s-1.12.yaml.template [hunk 1]
```diff
@@ -131,8 +131,8 @@ spec:
         {{- if or (not .Authentication.Aws.BackendMode) (contains "MountedFile" .Authentication.Aws.BackendMode) }}
         - --config=/etc/aws-iam-authenticator/config.yaml
         {{- end }}
-        {{- if or .Authentication.Aws.ClusterID (not (contains "MountedFile" .Authentication.Aws.BackendMode))}}
-        - --cluster-id={{ or .Authentication.Aws.ClusterID .MasterPublicName }}
+        {{- if or .Authentication.Aws.ClusterID (or (not .Authentication.Aws.BackendMode) (not (contains "MountedFile" .Authentication.Aws.BackendMode)))}}
```
Comment (johngmyers) [inline upup/models/cloudup/resources/addons/authentication.aws/k8s-1.12.yaml.template:134]: Can this be simplified?

Can this be simplified?

```suggestion
        {{- if or .Authentication.Aws.ClusterID (not .Authentication.Aws.BackendMode) (not (contains "MountedFile" .Authentication.Aws.BackendMode))}}
```

**Comment by fred-vogt** [comment]
@WarpRat - Excellent work my friend. Thanks for jumping on this so quickly.

I'm still building an integration around the newly available backendMode(s).

**Comment by WarpRat** [comment]
@fred-vogt - no problem, feel foolish that I missed this in the first place. I still need to make a minor update to bring the docs in line with the new default cluster id which @rifelpet suggested and the minor simplification that @johngmyers suggested above, but I should be able to handle that in an hour or two once I'm off work.

File: upup/models/cloudup/resources/addons/authentication.aws/k8s-1.12.yaml.template [hunk 2]
```diff
@@ -131,8 +131,8 @@ spec:
         {{- if or (not .Authentication.Aws.BackendMode) (contains "MountedFile" .Authentication.Aws.BackendMode) }}
         - --config=/etc/aws-iam-authenticator/config.yaml
         {{- end }}
-        {{- if or .Authentication.Aws.ClusterID (not (contains "MountedFile" .Authentication.Aws.BackendMode))}}
-        - --cluster-id={{ or .Authentication.Aws.ClusterID .MasterPublicName }}
+        {{- if or .Authentication.Aws.ClusterID (not .Authentication.Aws.BackendMode) (not (contains "MountedFile" .Authentication.Aws.BackendMode))}}
```
Comment (rifelpet) [inline upup/models/cloudup/resources/addons/authentication.aws/k8s-1.12.yaml.template:134]: I'm not sure i understand the logic change here.

The old logic would add `--cluster-id` if any of the following:
* ClusterID API field is set
* BackendMode does not contain MountedFile (including if BackendMode is not set)

This new logic will add `--cluster-id` if any of the following:
* ClusterID API field is set
* BackendMode API field is not set
* BackendMode API field does not contain MountedFile

Wouldn't we want to _not_ set --cluster-id if BackendMode is not set in order to preserve original behavior? 

```suggestion
        {{- if or .Authentication.Aws.ClusterID (and .Authentication.Aws.BackendMode (not (contains "MountedFile" .Authentication.Aws.BackendMode)))}}
```

The above would add `--cluster-id` if any of the following:
* ClusterID API field is set
* BackendMode API field is set and does not contain MountedFile

File: upup/models/cloudup/resources/addons/authentication.aws/k8s-1.12.yaml.template (hunk 1 above)
Comment (WarpRat) [reply upup/models/cloudup/resources/addons/authentication.aws/k8s-1.12.yaml.template]: Much clearer. Thanks!

File: upup/models/cloudup/resources/addons/authentication.aws/k8s-1.12.yaml.template (hunk 2 above)
Comment (WarpRat) [reply upup/models/cloudup/resources/addons/authentication.aws/k8s-1.12.yaml.template]: Yeah, you're right. In my local tests I saw the `--cluser-id` flag get replaced with the `--config` and didn't even notice the clusterID flag was just a few lines lower. I've tested all the different possibilities you mentioned here and with your suggested change everything is working as expected. Thanks for catching that.

**Comment by rifelpet** [comment]
👍 
/lgtm
/approve

**Pull request merged**